Readme.md
Rundeck JSON plugin
=======

version: 1.1

This is an example "Resource Format Plugin" for [RunDeck](http://rundeck.org).

It provides a parser and generator for resources in JSON format, using the
Google GSON library.

Dependencies:
-----

Copy the latest rundeck-core jar to the `build-dependencies` directory.

Build
-----

./gradlew

Artifact
-----

    build/libs/rundeck-json-plugin-1.1.jar

Install
-----

Copy the `rundeck-json-plugin-1.1.jar` to your `$RDECK_BASE/libext` directory.

You can now use the `resourcejson` format when configuring Resource Model Source
plugins, or exporting resources via the API.

For example, add a new File resource model source with a ".json" extension, or
a new URL resource model source for a URL that returns json.

When accessing resources via the API, use `?format=resourcejson` to retrieve the resources in this format.

Format
-----

The JSON format supported for nodes is:

~~~ {.json}
{
  "nodes": {
    "Venkman.local": {
      "tags": ["a","b","c"],
      "attributes": {
        "osFamily": "unix",
        "username": "greg",
        "osVersion": "10.7.1",
        "osArch": "x86_64",
        "description": "Rundeck server node",
        "hostname": "Venkman.local",
        "osName": "Mac OS X"
      }
    },
    "centos5": {
      "tags": ["d","b","c"],
      "attributes": {
        "osFamily": "unix",
        "username": "greg",
        "osVersion": "10.7.1",
        "osArch": "x86_64",
        "description": "Test node",
        "hostname": "centos5",
        "osName": "Mac OS X",
        "monkey":"banana",
        "eats-cheese":"yes"
      }
    }
  }
}
~~~

More Info
-----

More info about rundeck plugins:

* [Resource Format plugin usage](http://rundeck.org/docs/plugins-user-guide/resource-model-source-plugins.html#resource-model-document-formats)
* [Resource Format plugin development](http://rundeck.org/docs/developer/model-format-parser-and-generator-plugin.html)
* [Rundeck API - Resources for a project](http://rundeck.org/docs/api/index.html#updating-and-listing-resources-for-a-project)


CHANGELOG
# Changelog

## 1.1 - 2014-10-29

* fix: compatibility with Rundeck 2

## 1.0 - 2011-09-01

* Initial release

src/main/java/com/dtolabs/rundeck/plugin/resources/format/json/Node.java
/*
 * Copyright 2011 DTO Labs, Inc. (http://dtolabs.com)
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

/*
* Node.java
* 
* User: Greg Schueler <a href="mailto:greg@dtosolutions.com">greg@dtosolutions.com</a>
* Created: 8/8/11 11:27 AM
* 
*/
package com.dtolabs.rundeck.plugin.resources.format.json;

import java.util.*;

/**
* Node is ...
*
* @author Greg Schueler <a href="mailto:greg@dtosolutions.com">greg@dtosolutions.com</a>
*/
class Node {
    private String nodename;
    private Set<String> tags;
    private Map<String,String> attributes;

    public String getNodename() {
        return nodename;
    }

    public void setNodename(String nodename) {
        this.nodename = nodename;
    }

    public Set<String> getTags() {
        return tags;
    }

    public void setTags(Set<String> tags) {
        this.tags = tags;
    }

    public Map<String, String> getAttributes() {
        return attributes;
    }

    public void setAttributes(Map<String, String> attributes) {
        this.attributes = attributes;
    }
}


src/main/java/com/dtolabs/rundeck/plugin/resources/format/json/Nodes.java
/*
 * Copyright 2011 DTO Labs, Inc. (http://dtolabs.com)
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

/*
* Nodes.java
* 
* User: Greg Schueler <a href="mailto:greg@dtosolutions.com">greg@dtosolutions.com</a>
* Created: 8/8/11 11:28 AM
* 
*/
package com.dtolabs.rundeck.plugin.resources.format.json;

import java.util.*;

/**
* Nodes is ...
*
* @author Greg Schueler <a href="mailto:greg@dtosolutions.com">greg@dtosolutions.com</a>
*/
class Nodes {
    private TreeMap<String, Node> nodes;

    public Nodes() {
        this.nodes = new TreeMap<String, Node>();
    }

    public TreeMap<String, Node> getNodes() {
        return nodes;
    }

    public void setNodes(TreeMap<String, Node> nodes) {
        this.nodes = nodes;
    }
}


src/main/java/com/dtolabs/rundeck/plugin/resources/format/json/JsonResourceFormatGenerator.java
/*
 * Copyright 2011 DTO Solutions, Inc. (http://dtosolutions.com)
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

/*
* JsonResourceFormatParser.java
* 
* User: Greg Schueler <a href="mailto:greg@dtosolutions.com">greg@dtosolutions.com</a>
* Created: 3/31/11 3:14 PM
* 
*/
package com.dtolabs.rundeck.plugin.resources.format.json;


import com.dtolabs.rundeck.core.common.INodeEntry;
import com.dtolabs.rundeck.core.common.INodeSet;
import com.dtolabs.rundeck.core.common.NodeEntryImpl;
import com.dtolabs.rundeck.core.common.NodeSetImpl;
import com.dtolabs.rundeck.core.plugins.Plugin;
import com.dtolabs.rundeck.core.resources.format.ResourceFormatGenerator;
import com.dtolabs.rundeck.core.resources.format.ResourceFormatGeneratorException;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.io.*;
import java.util.*;

/**
 * JsonResourceFormatParser parsers JSON format into nodes data
 *
 * @author Greg Schueler <a href="mailto:greg@dtosolutions.com">greg@dtosolutions.com</a>
 */
@Plugin (name = "resourcejson", service = "ResourceFormatGenerator")
public class JsonResourceFormatGenerator implements ResourceFormatGenerator {
    public static final String SERVICE_PROVIDER_TYPE = "resourcejson";
    private static final Gson test = new GsonBuilder().setPrettyPrinting().create();

    public static final Set<String> EXTENSIONS = Collections.unmodifiableSet(new HashSet<String>(Arrays.asList(
        "json")));
    public static final List<String> MIME_TYPES = Collections.unmodifiableList(Arrays.asList(
        "application/json", "text/json"));

    public Set<String> getFileExtensions() {
        return EXTENSIONS;
    }

    public List<String> getMIMETypes() {
        return MIME_TYPES;
    }

    public void generateDocument(INodeSet nodeset, OutputStream stream) throws ResourceFormatGeneratorException {
        try {
            final NodeSetImpl nodeSet = new NodeSetImpl();
            nodeSet.putNodes(nodeset);

            final OutputStreamWriter outputStreamWriter = new OutputStreamWriter(stream, "UTF-8");
            try {
                outputStreamWriter.write(test.toJson(nodesFromNodeSet(nodeset)));
            } finally {
                outputStreamWriter.close();
            }
        } catch (IOException e) {
            throw new ResourceFormatGeneratorException(e);
        }
    }

    static Nodes nodesFromNodeSet(INodeSet nodeset) {
        Nodes nodes = new Nodes();
        final TreeMap<String, Node> map = new TreeMap<String, Node>();
        for (final INodeEntry iNodeEntry : nodeset.getNodes()) {

            String name = iNodeEntry.getNodename();
            final Node nodeEntry = new Node();
            nodeEntry.setNodename(name);
            if (null != iNodeEntry.getTags()) {
                nodeEntry.setTags(iNodeEntry.getTags());
            }
            if (null != iNodeEntry.getAttributes()) {
                nodeEntry.setAttributes(iNodeEntry.getAttributes());
            }
            map.put(name, nodeEntry);
        }
        nodes.setNodes(map);

        return nodes;
    }
}


src/main/java/com/dtolabs/rundeck/plugin/resources/format/json/JsonResourceFormatParser.java
/*
 * Copyright 2011 DTO Solutions, Inc. (http://dtosolutions.com)
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

/*
* JsonResourceFormatParser.java
* 
* User: Greg Schueler <a href="mailto:greg@dtosolutions.com">greg@dtosolutions.com</a>
* Created: 3/31/11 3:14 PM
* 
*/
package com.dtolabs.rundeck.plugin.resources.format.json;


import com.dtolabs.rundeck.core.common.INodeEntry;
import com.dtolabs.rundeck.core.common.INodeSet;
import com.dtolabs.rundeck.core.common.NodeEntryImpl;
import com.dtolabs.rundeck.core.common.NodeSetImpl;
import com.dtolabs.rundeck.core.plugins.Plugin;
import com.dtolabs.rundeck.core.resources.format.ResourceFormatParser;
import com.dtolabs.rundeck.core.resources.format.ResourceFormatParserException;
import com.google.gson.*;

import java.io.*;
import java.lang.reflect.Type;
import java.util.*;

/**
 * JsonResourceFormatParser parsers JSON format into nodes data
 *
 * @author Greg Schueler <a href="mailto:greg@dtosolutions.com">greg@dtosolutions.com</a>
 */
@Plugin (name = "resourcejson", service = "ResourceFormatParser")
public class JsonResourceFormatParser implements ResourceFormatParser {
    public static final String SERVICE_PROVIDER_TYPE = "resourcejson";
    private  final Gson gson;

    public static final Set<String> FILE_EXTENSIONS = Collections.unmodifiableSet(new HashSet<String>(Arrays.asList(
        "json")));

    public static final Set<String> MIME_TYPES = Collections.unmodifiableSet(new HashSet<String>(Arrays.asList(
        "application/json")));

    public JsonResourceFormatParser() {
        gson = new GsonBuilder().create();
    }

    public Set<String> getFileExtensions() {
        return FILE_EXTENSIONS;
    }

    public Set<String> getMIMETypes() {
        return MIME_TYPES;
    }

    public INodeSet parseDocument(File file) throws ResourceFormatParserException {
        try {
            final FileInputStream fileInputStream = new FileInputStream((file));
            try {
                final INodeSet iNodeSet = parseDocument(fileInputStream);
                System.err.println("node set: "+iNodeSet);
                return iNodeSet;
            } finally {
                fileInputStream.close();
            }
        } catch (IOException e) {
            throw new ResourceFormatParserException(e);
        }
    }

    public INodeSet parseDocument(InputStream inputStream) throws ResourceFormatParserException {
        final InputStreamReader json = new InputStreamReader(inputStream);
        final Nodes nodes;
        try {
            try {
                nodes = gson.fromJson(json, Nodes.class);
            } finally {
                json.close();
            }
        } catch (IOException e) {
            throw new ResourceFormatParserException(e);
        } catch (JsonIOException e){
            throw new ResourceFormatParserException(e);
        } catch (JsonSyntaxException e){
            throw new ResourceFormatParserException(e);
        }
        NodeSetImpl nodeset = nodeSetFromNodes(nodes);
        return nodeset;
    }

    static  NodeSetImpl nodeSetFromNodes(Nodes nodes) {
        NodeSetImpl nodeset = new NodeSetImpl();
        for (final Map.Entry<String, Node> entry : nodes.getNodes().entrySet()) {
            String name = entry.getKey();
            final Node value = entry.getValue();
            final NodeEntryImpl nodeEntry = new NodeEntryImpl();
            if (null != value.getAttributes()) {
                nodeEntry.setAttributes(value.getAttributes());
            }
            if(null!=value.getTags()){
                nodeEntry.setTags(value.getTags());
            }
            nodeEntry.setNodename(name);
            nodeset.putNode(nodeEntry);
        }
        return nodeset;
    }


}
